File: src/main/scala/com/exasol/cloudetl/kafka/KafkaConsumerQueryGenerator.scala
package com.exasol.cloudetl.kafka

import scala.collection.JavaConverters._

import com.exasol.ExaImportSpecification
import com.exasol.ExaMetadata

import com.typesafe.scalalogging.LazyLogging

/**
 * This object is referenced from the UDF script that is called from the
 * Exasol's {@code IMPORT FROM SCRIPT} SQL statement.
 *
 * It returns an SQL query that is run to import data from a Kafka
 * topic.
 */
object KafkaConsumerQueryGenerator extends LazyLogging {

  /**
   * An entry point for the Exasol IMPORT FROM SCRIPT user-defined
   * function.
   *
   * The generated SQL query calls metadata and import UDF scripts
   * internally. Additionally, the generated query runs in a single
   * transaction.
   */
  def generateSqlForImportSpec(
    metadata: ExaMetadata,
    importSpec: ExaImportSpecification
  ): String = {
    val kafkaProperties = KafkaConsumerProperties(importSpec.getParameters.asScala.toMap)
    val tableName = kafkaProperties.getTableName()
    val topic = kafkaProperties.getTopic()
    if (topic.contains(",")) {
      throw new IllegalArgumentException(
        "Please provide only a single topic name. Importing data " +
          "from multiple topics is not supported."
      )
    }
    logger.info(s"Generating a SQL query to import from '$topic'.")
    val kvPairs = kafkaProperties.mkString()
    val scriptSchema = metadata.getScriptSchema

    s"""SELECT
       |  $scriptSchema.KAFKA_IMPORT(
       |    '$kvPairs', partition_index, max_offset
       |)
       |FROM (
       |  SELECT $scriptSchema.KAFKA_METADATA(
       |    '$kvPairs', kafka_partition, kafka_offset
       |  ) FROM (
       |    SELECT kafka_partition, MAX(kafka_offset) AS kafka_offset
       |    FROM $tableName
       |    GROUP BY kafka_partition
       |    UNION ALL
       |    SELECT 0, -1
       |    FROM DUAL
       |    WHERE NOT EXISTS (SELECT * FROM $tableName LIMIT 2)
       |  )
       |)
       |GROUP BY
       |  partition_index;
       |""".stripMargin
  }

}

File: src/main/java/com/exasol/cloudetl/kafka/KafkaConsumerQueryGenerator.java
package com.exasol.cloudetl.kafka;

import com.exasol.ExaImportSpecification;
import com.exasol.ExaMetadata;
import com.exasol.errorreporting.ExaError;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class KafkaConsumerQueryGenerator {
    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaConsumerQueryGenerator.class);

    private KafkaConsumerQueryGenerator() {
    }

    public static String generateSqlForImportSpec(final ExaMetadata metadata, final ExaImportSpecification importSpec) {
        final KafkaConsumerProperties kafkaProperties = KafkaConsumerProperties.apply(importSpec.getParameters());
        final String tableName = kafkaProperties.getTableName();
        final String topic = kafkaProperties.getTopic();
        if (topic.contains(",")) {
            throw new IllegalArgumentException(ExaError.messageBuilder("E-KCE-10")
                    .message("Importing data from multiple topics is not supported.")
                    .mitigation("Please provide only a single topic name.")
                    .toString());
        }
        LOGGER.info("Generating a SQL query to import from '{}'.", topic);
        final String kvPairs = kafkaProperties.mkString();
        final String scriptSchema = metadata.getScriptSchema();
        return "SELECT\n"
                + "  " + scriptSchema + ".KAFKA_IMPORT(\n"
                + "    '" + kvPairs + "', partition_index, max_offset\n"
                + ")\n"
                + "FROM (\n"
                + "  SELECT " + scriptSchema + ".KAFKA_METADATA(\n"
                + "    '" + kvPairs + "', kafka_partition, kafka_offset\n"
                + "  ) FROM (\n"
                + "    SELECT kafka_partition, MAX(kafka_offset) AS kafka_offset\n"
                + "    FROM " + tableName + "\n"
                + "    GROUP BY kafka_partition\n"
                + "    UNION ALL\n"
                + "    SELECT 0, -1\n"
                + "    FROM DUAL\n"
                + "    WHERE NOT EXISTS (SELECT * FROM " + tableName + " LIMIT 2)\n"
                + "  )\n"
                + ")\n"
                + "GROUP BY\n"
                + "  partition_index;\n";
    }
}
